aspnetcore/security/key-vault-configuration.md

```diff
@@ -0,0 +1,135 @@
+
+## Application configuration
+You can explore the provider with the [sample application](https://github.com/aspnet/Docs/tree/master/aspnetcore/security/key-vault-configuration/sample). Once you've established a key vault and created a pair of secrets in the vault by [following the guidance below](#creating-key-vault-secrets-and-loading-configuration-values), the sample application will securely load the secret values into its configuration and display them in a webpage.
+
+The provider is added to the `ConfigurationBuilder` with the `AddAzureKeyVault()` extension. In the sample application, the extension uses three configuration values loaded from the *appsettings.json* file.
+
+App Setting | Description | Example
+--- | --- | ---
+`Vault` | Azure Key Vault name | contosovault
+`ClientId` | Azure Active Directory App Id | 627e911e-43cc-61d4-992e-12db9c81b413
+`ClientSecret` | Azure Active Directory App Key | g58K3dtg59o1Pa+e59v2Tx829w6VxTB2yv9sv/101di=
+
+[!code-csharp[Main](key-vault-configuration/sample/Startup.cs?name=snippet1&highlight=5,10-13)]
+
+`AddAzureKeyVault()` also provides an overload that accepts an implementation of `IKeyVaultSecretManager`, which allows you to control how key vault secrets are converted into configuration keys. For example, the interface can be implemented to load configuration values by environment, where you would prefix environment names to configuration secrets you've stored in the key vault. Key vault secrets `Development-ConnectionString` and `Production-ConnectionString` can be loaded from configuration as `ConnectionString`, and the application will take its configuration for the connection string from the secret that matches the application's environment. An example of this implementation is shown below.
+
+```csharp
+public class EnvironmentSecretManager : IKeyVaultSecretManager
+{
+    private readonly string _environmentPrefix;
+
+    public EnvironmentSecretManager(string environment)
+    {
+        _environmentPrefix = environment + "-";
+    }
+
+    public bool Load(SecretItem secret)
+    {
+        return secret.Identifier.Name.StartsWith(_environmentPrefix);
+    }
+
+    public string GetKey(SecretBundle secret)
+    {
+        return secret.SecretIdentifier.Name.Substring(_environmentPrefix.Length);
+    }
+}
+```
+
+```csharp
+builder.AddAzureKeyVault(
+    $"https://{config["Vault"]}.vault.azure.net/",
+    config["ClientId"],
+    config["ClientSecret"],
+    new EnvironmentSecretManager(config["ASPNETCORE_ENVIRONMENT"]));
+    
+Configuration = builder.Build();
+
+// Configuration["ConnectionString"] will be loaded from Development-ConnectionString
+// if the environment is Development and Production-ConnectionString if the environment 
+// is Production.
+```
+
+You can also provide your own `KeyVaultClient` implementation to `AddAzureKeyVault()`. For more information, see [KeyVaultClient Class](https://msdn.microsoft.com/en-us/library/microsoft.azure.keyvault.keyvaultclient.aspx). 
```

Comment:
Honestly I haven't done a ton of in-depth work looking at Key Vault, but the `KeyVaultClient` is the central component used to actually communicate with Azure Key Vault. It has a bunch of properties and configuration that the user may want to control. We only propagate a few of those through our API so we provide this API to allow the user to fully configure a `KeyVaultClient` as they wish and pass it along to us.